### ISSUE TITLE
REPL/interactive mode for multiple queries

### ISSUE THREAD

**Issue by Suguivy** [marcua/ayb#275]
Right now, to do multiple queries in `ayb`, one must write `ayb query <database> <query>` for each query they want to execute in the database. Adding a REPL/interactive mode for multiple queries could help to execute multiple queries easier.

For example, instead of doing this:

```sh
sugui@yukari:~$ ayb client query sugui/test-from-cli "create table numbers(spelling varchar, num integer);"

Rows: 0
sugui@yukari:~$ ayb client query sugui/test-from-cli "insert into numbers values ('seven', 7)"

Rows: 0
sugui@yukari:~$ ayb client query sugui/test-from-cli "select * from numbers"
 spelling | num 
----------+-----
 seven    | 7 

Rows: 1
```

It could exist something like a `client interactive` option to enter an interactive mode where you can simply start to write queries into the same database.
```sh
sugui@yukari:~$ ayb client interactive sugui/test-from-cli 
sugui@sugui/test-from-cli> create table numbers(spelling varchar, num integer);

Rows: 0
sugui@sugui/test-from-cli> insert into numbers values ('seven', 7);

Rows: 0
sugui@sugui/test-from-cli> select * from numbers;
 spelling | num 
----------+-----
 seven    | 7 

Rows: 1
sugui@sugui/test-from-cli> ^C
sugui@yukari:~$
```

This is an improvement I told @sofiaritz about. She told me I could open an issue for that, so here it is :)

**Comment by marcua** [comment]
Thank you for this excellent idea, @Suguivy! :) I would love to get this feature into `ayb`. Your proposal looks great to me. Two thoughts on this implementation
* It looks like we can rely on [rustyline](https://crates.io/crates/rustyline) do do the heavy lifting of reading a line of input with support for history, hotkeys, etc.
* Can you or @sofiaritz think of any reason why we need to give it its own CLI command? For example, do we want `ayb client interactive sugui/test-from-cli ` with an explicit `interactive` command, or should it be `ayb client query sugui/test-from-cli`, and if there's no query at the end of this command, it implicitly goes into interactive mode?

I'm quite excited about this idea, and would love to either help you contribute it if you wanted to write the code, or to write it and get your feedback on an early implementation. (And no problem if you wanted to share the idea but not more than that.) Do you have a preference?

**Comment by Suguivy** [comment]
Thank you for the response @marcua! I will reply now to the two points you highlighted:

- I think that the [rustyline](https://crates.io/crates/rustyline) library you linked to, is a good choice to automatically manage things like hotkeys and history. Another important thing is, do we want to store the history in a file to use it on proper executions? In my opinion it's a good idea, to allow revisiting past queries.

- I think that it can be a good idea to use the same `ayb client query <database>` command without any query to automatically enter the interactive mode. We don't really need any special command for that, but I think we should document it in the description of the command when using the `--help` flag.

And yes, I'd like to write the feature! It's been a while since I don't code in Rust and maybe my experience in the language is not so broad, but I'll try my best. I don't have many time because of work but I think it's not so hard to implement, so I'd be pleased to contribute. 😊

**Comment by sofiaritz** [comment]
I agree with @Suguivy points, although the persistent query history may be complex to implement (should the history be global or db-specific? how do we allow users to disable it? what about logging into a different instance, how should history react to that? etc).

Re: contributing. Feel free to ask me any questions you may have or any problems that may arise while implementing this feature in Matrix :)
No pressure though :p

**Comment by marcua** [comment]
Hooray @Suguivy and thank you @sofiaritz! It sounds like `rustyline,` `ayb client query <database>` with no query, and a `--help` update is what we'll need. To @sofiaritz 's point, supporting saved history is helpful but likely complicated experience-wise, so perhaps in the v1 we can benefit from rustyline's history within a session, but not across sessions (no need to save to a file).

It's wonderful you'd like to contribute code. :) If you end up not having enough time, etc., just let us know and we can jump in to help, since this feature is a great idea! In terms of getting started:
* All/almost all of the changes are likely to be in https://github.com/marcua/ayb/blob/main/src/bin/ayb.rs
* [Around this part of the code](https://github.com/marcua/ayb/blob/5aa9f8ac3c898434f30234e7ecf84079306da048/src/bin/ayb.rs#L97) you can make `query` be not required, and update the `--help` text.
* [This region of the code](https://github.com/marcua/ayb/blob/5aa9f8ac3c898434f30234e7ecf84079306da048/src/bin/ayb.rs#L363) is where you can see if `query` is missing, and if so, add the `rustyline` loop.

I'm grateful for your submitting this issue, and excited at the prospect of your first contribution. Thank you so much, and thank you to @sofiaritz for being so welcoming! :)

**Comment by marcua** [comment]
Thank you again for this excellent idea and for your code contribution @Suguivy! I've submitted and merged #292, which is a follow-on to the code. I ran into similar issues with you on error types, so I didn't implement the `?` recommendations I initially made. The PR is mostly `cargo fmt` and `cargo clippy` with some additional clarifications of error messages and an update to the README since we can now show off our cool new interactive mode.

I hope you find `ayb` useful, and that you feel welcome to contribute to the project in the future. Your feedback so far has been wonderful, and I look forward to more of it! :)
